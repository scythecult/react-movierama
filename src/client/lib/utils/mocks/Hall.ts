import type { SeatNode } from './SeatNode';

type HallOptions = {
  Seat: typeof SeatNode;
  hallPlanMatrix: number[][];
  seatWidth: number;
  seatHeight: number;
  seatOffset: number;
};

export class Hall {
  #hallPlanMatrix: number[][];
  #seatWidth: number;
  #seatHeight: number;
  #seatOffset: number;
  #canvasWidth = 0;
  #canvasHeight = 0;
  #finalRowLength = 0;
  #Seat: typeof SeatNode;
  #hall: SeatNode[] = [];

  constructor(options: HallOptions) {
    this.#Seat = options.Seat;
    this.#hallPlanMatrix = options.hallPlanMatrix;
    this.#seatWidth = options.seatWidth;
    this.#seatHeight = options.seatHeight;
    this.#seatOffset = options.seatOffset;
  }

  createSeats() {
    for (let i = 0; i < this.#hallPlanMatrix.length; i++) {
      const currentRowLength = this.#hallPlanMatrix[i].length - 1;

      for (let j = 0; j < this.#hallPlanMatrix[i].length; j++) {
        const seatMatrixValue = this.#hallPlanMatrix[i][j];
        const isGap = !seatMatrixValue;

        if (this.#finalRowLength < currentRowLength) {
          this.#finalRowLength = currentRowLength;
        }

        this.#canvasWidth = this.#finalRowLength * (this.#seatWidth + this.#seatOffset) + this.#seatWidth;
        this.#canvasHeight = i * (this.#seatHeight + this.#seatOffset) + this.#seatHeight;

        if (!isGap) {
          const seatXOffset = isGap
            ? this.#seatWidth + this.#seatOffset + this.#seatWidth / 2
            : this.#seatWidth + this.#seatOffset;
          const x = j * seatXOffset;
          const y = i * (this.#seatHeight + this.#seatOffset);

          this.#hall.push(
            new this.#Seat({
              id: x + y,
              x,
              y,
              place: j + 1,
              type: seatMatrixValue,
              row: i + 1,
            }),
          );
        }
      }
    }

    return this.#hall;
  }

  getCanvasSize() {
    if (!this.#canvasWidth || !this.#canvasHeight) {
      this.createSeats();
    }

    return {
      width: this.#canvasWidth,
      height: this.#canvasHeight,
    };
  }
}
