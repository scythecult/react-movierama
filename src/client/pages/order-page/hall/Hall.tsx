import type { PropsWithChildren } from 'react';
import styles from './styles.module.css';

export type HallProps = PropsWithChildren;

export const Hall = (props: HallProps) => {
  const { children } = props;

  return (
    <div className={styles.hall}>
      <div className={styles.hallScreen}></div>
      {children}
    </div>
  );
};
