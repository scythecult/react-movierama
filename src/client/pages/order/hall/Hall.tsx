import type { PropsWithChildren } from 'react';
import styles from './styles.module.css';

export const Hall = (props: PropsWithChildren) => {
  const { children } = props;

  return (
    <div className={styles.hall}>
      <div className={styles.hallScreen}></div>
      {children}
    </div>
  );
};
