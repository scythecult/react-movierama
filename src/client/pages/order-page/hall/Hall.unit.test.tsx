import { render } from '@testing-library/react';
import { Hall } from './Hall';

describe('Hall', () => {
  test('should correspond default layout', () => {
    const result = render(<Hall />);

    expect(result.container).toMatchSnapshot();
  });

  test('should support the "children" prop', () => {
    const result = render(<Hall>Children</Hall>);

    expect(result.container).toMatchSnapshot();
  });
});
