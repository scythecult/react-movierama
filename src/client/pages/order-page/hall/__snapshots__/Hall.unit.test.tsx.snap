// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Hall > should correspond default layout 1`] = `
<div>
  <div
    class="_hall_23b2aa"
    data-test-id="hall"
  >
    <div
      class="_hallScreen_23b2aa"
      data-test-id="hall-screen"
    />
  </div>
</div>
`;

exports[`Hall > should support the "children" prop 1`] = `
<div>
  <div
    class="_hall_23b2aa"
    data-test-id="hall"
  >
    <div
      class="_hallScreen_23b2aa"
      data-test-id="hall-screen"
    />
    Children
  </div>
</div>
`;
